source/fundamentals/crud/compound-operations.txt

```diff
@@ -0,0 +1,293 @@
+- :java-docs:`FindOneAndReplaceOptions <apidocs/mongodb-driver-core/com/mongodb/client/model/FindOneAndReplaceOptions.html>`
+
+Find and Delete One Document
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To find and delete a document, use the ``findOneAndDelete()`` method of the
+``MongoCollection`` class.
+
+The following example uses the ``findOneAndDelete()`` method to find and
+delete the document with the largest value in the ``_id`` field.
+
+The example uses a ``FindOneAndDeleteOptions`` instance to specify a
+descending sort on the ``_id`` field.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/CompoundOperatorsIndividualExamples.java
+   :language: java
+   :dedent:
+   :start-after: start findOneAndDelete-example
+   :end-before: end findOneAndDelete-example
+
+The output of the code above should look like this:
+
+.. code-block:: json
+
+   {"_id": 2, "food": "pear", "color": "yellow"}
+
+For more information on the ``Sorts`` class, see our
+:doc:`guide on the Sorts builder </fundamentals/builders/sort/>`.
+
+For more information on the methods and classes in this section, see the
+following API documentation pages:
+
+- :java-docs:`findOneAndDelete() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#findOneAndDelete(org.bson.conversions.Bson,com.mongodb.client.model.FindOneAndDeleteOptions)>`
+- :java-docs:`FindOneAndDeleteOptions <apidocs/mongodb-driver-core/com/mongodb/client/model/FindOneAndDeleteOptions.html>`
+
+Avoiding A Race Condition
+-------------------------
+
+In this section, we will go through an example that shows how we can
+avoid a race condition with compound operations.
+
+For this example, let's imagine that we run a hotel with one room and that we
+have a small Java program to help us checkout this room to a guest.  
+
+We represent our one room with this document in MongoDB. 
+
+.. code-block:: json
+
+    {"_id": 1, "guest": null, "room": "Blue Room", "reserved": false}
+
+`The full code for this example is available on Github here. <https://github.com/mongodb/docs-java/blob/master/source/includes/fundamentals/code-snippets/CompoundOperators.java>`__
+
+Example With Race Condition
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let's say our app uses this unsafe ``bookARoom`` method to checkout our room to
+a guest.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/CompoundOperators.java
+   :language: java
+   :dedent:
+   :emphasize-lines: 3,11
+   :start-after: start the-unsafe-book-a-room
+   :end-before: end the-unsafe-book-a-room
+
+Imagine two separate guests, Jan and Pat, try to book the room with this method
+at the same time.
+
+Jan sees this output:
+
+.. code-block:: none
+
+   You got the Blue Room Jan
+
+And Pat sees this output:
+
+.. code-block:: none
+
+   You got the Blue Room Pat
```

Comment:
Thanks for the note. I would like to keep the guest names for the following reasons:

- As concurrency is something that is often hard for people to understand, I want to try and be as clear as possible when things are happening on different machines, and I think having one message could allow people to misread this and think that only one machine/process is involved. 
- The runnable code sample will be hard to understand without guest names included. Currently the java program can be run, and it will be hard to follow without guest names.